deep_qa/data/instances/text_classification/frame_instance.py

```diff
@@ -0,0 +1,198 @@
+    @overrides
+    def read_from_line(cls, line: str):
+        """
+        Reads a FrameInstance from a line.  The format is:
+        frame represented as list of <role:role value phrase of maxlen 5> TAB <label>
+        e.g., from
+        event:plant absorb water###participant:water###agent:plant###finalloc:soil
+              to
+        ["plant", "unk", "unk", "unk", "unk", "plant absorb water",
+          "soil", "unk", "unk", "unk", "unk", "unk",
+          "unk", "unk", "unk", "unk", "unk", "unk",
+          "unk", "unk", "unk", "unk", "unk", "unk",
+          "unk", "unk", "water"]
+        Provides ordering (input can be composed of slots in arbitrary order)
+        and sparseness flexibility (only a few slots can be mentioned in the input).
+        """
+        # Extract the query slot name and expected value
+        # e.g. from, participant:water, extract the expected slot value "water"
+        unpacked_input = cls.unpack_input(line)
+        given_slots = cls.given_slots_from(unpacked_input['content'])
+        query_slot = cls.query_slot_from(unpacked_input['query'], given_slots)
+        padded_slots = cls.padded_slots_from(given_slots, query_slot['name'])
+        return cls(padded_slots, phrase_in_queried_slot=query_slot['val'])
+
+    @overrides
+    def to_indexed_instance(self, data_indexer: DataIndexer):
+        # A phrase in a slot, is converted from list of words to list of wordids.
+        # This is repeated for every slot, hence a list of list of wordids/integers.
+        indices_slotvals = [self._index_text(phrase, data_indexer) for phrase in self.text]
+        # The label is a phrase, and is converted from list of words to list of wordids.
+        indices_label = self._index_text(self.label, data_indexer)
+        return IndexedFrameInstance(indices_slotvals, indices_label)
+
+
+class IndexedFrameInstance(IndexedInstance):
+    """
+    Ensures that a phrase in every slot, and the label (also a phrase) are padded to be of a fixed maxlen.
+    Max length of a phrase is 6 (configurable), pad phrases with fewer words; if it exceeds 6 then truncate.
+    """
+    def __init__(self, word_indices: List[List[int]], label):
+        """
+        :param word_indices: One list of ints make up a slotvalue because a slotvalue is a phrase,
+                             and so every word of the phrase is identified with an int id.
+        :param label: phrase, hence a list of ints.
+        """
+        super(IndexedFrameInstance, self).__init__(label)
+        self.word_indices = word_indices
+
+    @classmethod
+    @overrides
+    def empty_instance(cls):
+        return IndexedFrameInstance([], label=None)
+
+    @overrides
+    def get_padding_lengths(self) -> Dict[str, int]:
+        # The parent class somehow understands the semantics of these dictionary keys.
+        # would be better if these keys were more explicit, e.g., as enumerations.
+        return {'num_sentence_words': MAX_PHRASE_LEN}
```

Comment:
Yeah, there's no need to hard-code a length here, and that parameter is better specified in the model you build, anyway.  This will actually break the code we have that tries to be smart about letting you swap out word representations.  It looks like the tuple instance code that Mark linked to also will break the word representation stuff that we have.  Instead, what you should do is something like [this](https://github.com/allenai/deep_qa/blob/dd6f20cd0fde902cc54f58baff29f355a356c3cd/deep_qa/data/instances/entailment/sentence_pair_instance.py#L66-L71).

@DeNeutoy, I think we need way better documentation around how the whole data pipeline works.  I think it does some really nice things, but people have a hard time understanding it and how to use it correctly.